fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -29,38 +29,80 @@
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 
 import static com.apple.foundationdb.record.lucene.LuceneIndexOptions.INDEX_PARTITION_BY_TIMESTAMP;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
 import static com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer.Waits.WAIT_LOAD_LUCENE_PARTITION_METADATA;
 
 /**
  * Manage partitioning info for a <b>logical</b>, partitioned lucene index, in which each partition is a separate physical lucene index.
  */
 @API(API.Status.EXPERIMENTAL)
 public class LucenePartitioner {
+    private static final ConcurrentHashMap<String, KeyExpression> partitioningKeyExpressionCache = new ConcurrentHashMap<>();
     public static final int PARTITION_META_SUBSPACE = 0;
     public static final int PARTITION_DATA_SUBSPACE = 1;
     private final IndexMaintainerState state;
     private final boolean partitioningEnabled;
     private final String partitionTimestampFieldName;
+    private final KeyExpression partitioningKeyExpression;
 
     public LucenePartitioner(@Nonnull IndexMaintainerState state) {
         this.state = state;
         partitionTimestampFieldName = state.index.getOption(INDEX_PARTITION_BY_TIMESTAMP);
         this.partitioningEnabled = partitionTimestampFieldName != null;
+        if (partitioningEnabled && (partitionTimestampFieldName.isEmpty() || partitionTimestampFieldName.isBlank())) {
+            throw new RecordCoreArgumentException("Invalid partition timestamp field name", LogMessageKeys.FIELD_NAME, partitionTimestampFieldName);
+        }
+        this.partitioningKeyExpression = makePartitioningKeyExpression(partitionTimestampFieldName);
+    }
+
+    /**
+     * make (and cache) a key expression from the partitioning field name.
+     *
+     * @param partitionTimestampFieldName partitioning field name
+     * @return key expression
+     */
+    @Nullable
+    private KeyExpression makePartitioningKeyExpression(@Nullable final String partitionTimestampFieldName) {
+        if (partitionTimestampFieldName == null) {
+            return null;
+        }
+
+        return partitioningKeyExpressionCache.computeIfAbsent(partitionTimestampFieldName, k -> {
+            // here, partitionTimestampFieldName is not null/empty/blank
+            String[] nameComponents = k.split("\\.");
+
+            // nameComponents.length >= 1
+            if (nameComponents.length == 1) {
+                // no nesting
+                return field(nameComponents[0]);
+            }
+            // nameComponents.length >= 2
+            List<KeyExpression> fields = Arrays.stream(nameComponents).map(Key.Expressions::field).collect(Collectors.toList());
+            for (int i = fields.size() - 1; i > 0; i--) {
+                fields.set(i - 1, ((FieldKeyExpression) fields.get(i - 1)).nest(fields.get(i)));
+            }
+            return fields.get(0);
```

Comment:
`Key.Expressions::field` returns a `FieldKeyExpression`, so if you change the type of fields to `List<FieldKeyExpression>` I think you can remove the cast.

Reply:
`.nest()` returns a `NestingKeyExpression` which is not a `FieldKeyExpression`, so the list needs to be of the base type.